manager/metrics/collector.go

```diff
@@ -0,0 +1,101 @@
+package metrics
+
+import (
+	"context"
+
+	"strings"
+
+	metrics "github.com/docker/go-metrics"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/state/store"
+)
+
+var (
+	nodesMetric metrics.LabeledGauge
+)
+
+func init() {
+	ns := metrics.NewNamespace("swarmkit", "", nil)
+	nodesMetric = ns.NewLabeledGauge("nodes", "The number of nodes", metrics.Total, "state")
+	for _, state := range api.NodeStatus_State_name {
+		nodesMetric.WithValues(strings.ToLower(state)).Set(0)
+	}
+	metrics.Register(ns)
+}
+
+// Collector collects swarmkit metrics
+type Collector struct {
+	store *store.MemoryStore
+
+	// stopChan signals to the state machine to stop running.
+	stopChan chan struct{}
+	// doneChan is closed when the state machine terminates.
+	doneChan chan struct{}
+
+	nodesByState map[string]api.NodeStatus_State
+}
+
+// NewCollector creates a new metrics collector
+func NewCollector(store *store.MemoryStore) *Collector {
+	return &Collector{
+		store:        store,
+		stopChan:     make(chan struct{}),
+		doneChan:     make(chan struct{}),
+		nodesByState: make(map[string]api.NodeStatus_State),
+	}
+}
+
+func (c *Collector) updateNodeState(n *api.Node, remove bool) {
+	if prevState, ok := c.nodesByState[n.ID]; ok {
+		nodesMetric.WithValues(strings.ToLower(prevState.String())).Dec(1)
+	}
+	if remove {
+		delete(c.nodesByState, n.ID)
+		return
+	}
+	newState := n.Status.State
+	nodesMetric.WithValues(strings.ToLower(newState.String())).Inc(1)
+	c.nodesByState[n.ID] = newState
+}
+
+// Run contains the collector event loop
+func (c *Collector) Run(ctx context.Context) error {
+	defer close(c.doneChan)
+
+	watcher, cancel, err := store.ViewAndWatch(c.store, func(readTx store.ReadTx) error {
+		nodes, err := store.FindNodes(readTx, store.All)
+		if err != nil {
+			return err
+		}
+		for _, node := range nodes {
+			c.updateNodeState(node, false)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	for {
+		select {
+		case event := <-watcher:
+			switch v := event.(type) {
+			case api.EventCreateNode:
+				c.updateNodeState(v.Node, false)
+			case api.EventUpdateNode:
+				c.updateNodeState(v.Node, false)
+			case api.EventDeleteNode:
+				c.updateNodeState(v.Node, true)
```

Comment:
Maybe it would be slightly clearer to have separate `updateNodeState` and `removeNodeState` methods.